### ISSUE TITLE
Incorrect unification for members of number class

### ISSUE THREAD

**Issue by TobyShaw** [elm-lang/elm-compiler#1480]
Tested in 0.17.1 and the Try Elm page (as of 7th Sept 2016), this bug is browser-independent.

This code exemplifies the error as concise as I could get it:

```
fakeInt : Int
fakeInt = unsafeCast 0.5

fakeFloat : Float
fakeFloat = unsafeCast fakeInt

asTypeOf : a -> a -> a
asTypeOf a b = a

unsafeCast : number -> number'
unsafeCast n  =
    let val = n `asTypeOf` 0
    in val
```

The 2nd argument of asTypeOf is necessary to recreate the bug. The bug arises from the unification of the 1st and 2nd argument of asTypeOf.

Additionally,

Replacing the "let x = y in x" with just "y" causes the error to be caught whenever the function is directly unified with a Float (ie if I ever try to apply unsafeCast to a known float,). It is not caught with this replacement if there is no other unification attempt. 

This bug may not be specific to number, but seems to be discovered faster when replacing number with another typeclass.

**Comment by JoeyEremondi** [comment]
Just to clarify, the bug is that this should *not* typecheck, but it passes the checker?

This lets you do bad stuff. For example, the following works:
    
```elm
main = 
  text <| toString <| List.take (unsafeCast 1.5) [1,2,3] -- prints [1,2]
```

Similarly, the values aren't rounded, it's literally letting us treat floats as ints.

```elm
main = 
  text <| toString <| (unsafeCast 5.5) % 3 -- prints 2.5
```

**Comment by TobyShaw** [comment]
It should not type check but does, you are correct.

Also I realised just now that it works from number -> number', and isn't specific to Ints, I will update the issue.

**Comment by jvoigtlaender** [comment]
Has one of you checked #1373 does not already cover this thing?

**Comment by TobyShaw** [comment]
Ahhh, actually #1268 covers a very similar bug, they're likely to have the same cause.
